# INFO 212: Data Science Programming 1
## CCI at Drexel University


## Lab: Aggregation


In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


## What to do:
- Answer each question individually.
- Comment on each single statement.
- Each missing comment will be penalized by 1 point.
- Verify the code by showing meaningful results.

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## Exercise 1
For the following analysis tasks. feel free to use visualization to illustrate the results.

1. Download and upload the train.csv. Create a DataFrame df from the data set. Show information of the DataFrame.

In [5]:
df = pd.read_csv('/content/drive/MyDrive/info212/train.csv') #Making a dataFrame using the pandas library.
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

2. The Survived column has binary values 1 and 0. Compute the average Survived rate grouped by Pclass in ascending order. Discuss insights you can draw from the results.

In [11]:
df.groupby('Pclass')['Survived'].mean() #Did a groupby on Pclass to get this mean Survival rate(mean).

Pclass
1    0.629630
2    0.472826
3    0.242363
Name: Survived, dtype: float64

Passengers in 1st class had the highest survival rate (62.96%), followed by 2nd class (47.28%) and 3rd class (24.24%). This indicates that higher-class passengers had better access to lifeboats or were prioritized during evacuation.

3. The Survived column has binary values 1 and 0. Compute the average Survived rate grouped by Pclass and Cabin. Show the results in Survived rate ascending order within each Pclass group. Discuss insights you can draw from the results.

In [13]:
df.groupby(['Pclass', 'Cabin'])['Survived'].mean() #Did a groupby on Pclass and Cabin to get this mean Survival rate(mean).

Pclass  Cabin
1       A10      0.0
        A14      0.0
        A16      1.0
        A19      0.0
        A20      1.0
                ... 
3       F E69    1.0
        F G63    0.0
        F G73    0.0
        F38      0.0
        G6       0.5
Name: Survived, Length: 147, dtype: float64

Within each Pclass, survival rates varied by cabin, indicating that cabin location likely impacted access to lifeboats (e.g., A16 in 1st class had 100% survival, while A10 had 0%). Overall, lower-class cabins had lower survival rates, emphasizing a disadvantage in evacuation access.








4. Define a function taking a DataFrame, Age column, and index as paramenter. The function categorizes Ages into Young, Middle_aged, and Senior category. Compute the Survived rate for each age group in descending order. Discuss insights you can draw from the results.

In [17]:
def age_survival_rate(df, age_column="Age"):
    # Define age categories based on ranges
    categories = ['Young', 'Middle_aged', 'Senior']

    # Create 'AgeGroup' column with categorized age groups
    df['AgeGroup'] = pd.cut(df[age_column], bins=[0, 29, 59, float('inf')], labels=categories, right=False)

    # Compute survival rate for each age group and sort in descending order
    survival_rate = df.groupby('AgeGroup', observed=True)['Survived'].mean().sort_values(ascending=False)
    return survival_rate


survival_rates_by_age = age_survival_rate(df)
print(survival_rates_by_age)

AgeGroup
Middle_aged    0.419255
Young          0.406593
Senior         0.250000
Name: Survived, dtype: float64


Middle-aged passengers had the highest survival rate (41.93%), followed closely by young passengers (40.66%), while seniors had the lowest (25%), likely due to mobility challenges. This suggests younger and middle-aged groups had an advantage in survival odds.








5. Map the Embarked cities as 'S' to 'Southampton', 'C' to 'Cherbourg', 'Q' to 'Queenstown'. Compute the Survived rate by each embarked city using the mapping. Discuss insights you can draw from the results.

In [18]:
# Map 'Embarked' codes to city names
df['Embarked'] = df['Embarked'].map({'S': 'Southampton', 'C': 'Cherbourg', 'Q': 'Queenstown'})

# Compute the survival rate for each embarked city
survival_rate_by_city = df.groupby('Embarked')['Survived'].mean()
print(survival_rate_by_city)


Embarked
Cherbourg      0.553571
Queenstown     0.389610
Southampton    0.336957
Name: Survived, dtype: float64


Cherbourg passengers had the highest survival rate (55.36%) maybe because of the presence of more 1st Pclass passengers. Lower survival rates for Queenstown (38.96%) and Southampton (33.70%) suggest more 3rd Pclass passengers.








6. Define a function that computes the peak difference between the maximum and minimum values in an array. Group the data by Pclass and compute the peak difference of Ages for each Pclass group. Show the maximum, minimum, and the difference between max and min. Discuss insights you can draw from the results.

In [22]:
# Define the function to compute peak difference
def peak_difference(arr):
    max_val = np.max(arr)
    min_val = np.min(arr)
    difference = max_val - min_val
    return max_val, min_val, difference

# Applied a lambda function on peak_difference function to where x is the table with each Pclass grouped to their age with droped the null values.
age_peak_difference_by_pclass = df.groupby('Pclass')['Age'].apply(lambda x: peak_difference(x.dropna()))

print(age_peak_difference_by_pclass)

Pclass
1    (80.0, 0.92, 79.08)
2    (70.0, 0.67, 69.33)
3    (74.0, 0.42, 73.58)
Name: Age, dtype: object


First class has the widest age range (0.92 to 79.08), while third class also shows a broad range (0.42 to 73.58), indicating diverse age groups across classes. The wide age span in third class suggests many families or younger passengers, while first class likely includes older, wealthier people.






